flips/20211208-cadence-remove-optional-reference.md

```diff
@@ -0,0 +1,116 @@
+
+| Status        | (Approved)       |
+:-------------- |:---------------------------------------------------- |
+| **FLIP #**    | [722](https://github.com/onflow/flow/pull/722) |
+| **Author(s)** | Daniel Sainati (daniel.sainati@dapperlabs.com)       |
+| **Sponsor**   | Daniel Sainati (daniel.sainati@dapperlabs.com)       |
+| **Updated**   | 2021-12-09                                           |
+
+## Objective
+
+This removes the ability to create a reference to an optional value in Cadence by 
+creating a reference to the result of an indexed access. This was previously the 
+only way to obtain a reference to an optional, and as such this change effectively 
+removes these references from the language entirely. Instead, it allows users to 
+make the type on the right-hand side of an `as` expression an optional, and 
+allows users to create optionally-typed references in place of references
+to optional values. 
+
+## Motivation
+
+As detailed in https://github.com/onflow/cadence/issues/747, it is possible to 
+crash execution with a failed dereference that is not caught by the typechecker
+by creating a reference to the result of an indexed access, and then accessing
+a field on that reference. This occurs because while references to optional 
+values are normally disallowed, references to indexed accesses are special cased
+in the checker. In order to proceed with a stable release of Cadence, this feature
+needs to be removed. 
+
+## User Benefit
+
+This removes an unsound behavior from the type checker, preventing an error
+from occurring at runtime. 
+
+## Design Proposal
+
+When creating a reference value, the Cadence type checker includes a
+check to ensure that the type of the referenced value is not optional.
+However, prior to this change, there was a special case included that would
+"unwrap" one layer of an optional obtained from an indexed access, effectively
+allowing people to create optional references when the optional was the 
+result of an indexed access. 
+
+This worked without issue when the indexed access succeeded, but if the
+access would have returned `nil`, users were able to access the fields of the
+non-`nil` case of the optional value, which was unsafe and caused a crash at
+runtime.
+
+This FLIP proposes that instead of creating a reference to the result of the 
+access and assuming that the access succeeded, instead the reference itself 
+will be optional. In the case where the access succeeded, the reference will
+point to the value that was returned, and when it failed, the reference itself
+will be `nil` (as opposed to a reference to `nil`). This will be denoted in the
+program itself by marking the type on the right-hand side an optional.
+
+So, the following code would now be a type error, where it was not before:
+
+```cadence
+pub struct S {
+    pub let foo : Number
+    init() {
+        self.foo = 0
+    }
+}
+let dict: {String: S} = {}
+let s = &dict[""] as &S? // s has type &S?
```

Comment:
Maybe document here that `&S?` reads `(&S)?`, i.e. the result of the reference expression is going to be an optional, and not `&(S?)` (the result is a reference to an optional)